### PR TITLE
Ensure old ssl certs not copied in nginx role

### LINKED ISSUE #145: SSL certificates get overwritten in nginx role
### Short Description

The following task in the `mirsg.infrastructure.nginx` role copies SSL/TLS certificates into place:

https://github.com/UCL-MIRSG/ansible-collection-infra/blob/18bd51788d8d92be28d3ab8a5eafe879f8692d46/roles/nginx/tasks/main.yml#L48-L62

However, new certificates requested from UCL CS TSG are manually copied into place by TSG staff. Unless the new certificates are copied back to the location defined by `nginx_server_cert_cache` running the `nginx` role will copy old (likely expired) certificates into place.

### Evidence/Steps to Reproduce

_No response_

### Acceptance Criteria/Expected Behaviour

_No response_

### Details

_No response_

### Resolution

_No response_

### PR DESCRIPTION
**Pull request by ruaridhg**
Fixes #145 

Checks if ssl certificates exist on server and in cache and copies ssl certificates based on existence and expiry dates:

- Copy from cache to server if ssl cert doesn't exist on server
- Copy from server to cache if ssl cert doesn't exist in cache
- Copy from cache to server if ssl cert on server expires sooner
- Copy from server to cache if ssl cert in cache expires sooner

### CODE REVIEW

**Review by drmatthews — CHANGES_REQUESTED** [review]
Unfortunately I don't think this is going to solve our problem on its own. If CS put the new cert in the path defined by `nginx_ssl_cert_file`, while the one in the cache is the old cert, the checksums will be different and the new cert will get overwritten (i.e. it still requires someone to copy the newly placed certs into the cache location).

**Comment by ruaridhg** [comment]
> Unfortunately I don't think this is going to solve our problem on its own. If CS put the new cert in the path defined by `nginx_ssl_cert_file`, while the one in the cache is the old cert, the checksums will be different and the new cert will get overwritten (i.e. it still requires someone to copy the newly placed certs into the cache location).

Ah, yep that makes sense. Okay, I'll work on this more on Monday.

**Review by p-j-smith — CHANGES_REQUESTED** [review]
to add to @drmatthews point, the `x509_certificate_info_module` will probably be useful here -  you could check and compare the expiry dates ([`not after`](https://docs.ansible.com/ansible/latest/collections/community/crypto/x509_certificate_info_module.html#return-not_after)) of the certs. It might also be worth setting `backup: true` when copying the cert and key from cache

**Review by p-j-smith — CHANGES_REQUESTED** [review]
looking good! There are quite a few tasks that run `when: nginx_use_ssl` - perhaps we could put them into a separate file and include the file when SSL is enabled?

**Review by p-j-smith — COMMENTED** [review]
@ruaridhg I've pushed some changes to:
- check certificate expiry date in the nginx role before copying to/from cache
- added variable to `ssl_certificates` role to set the expiry date (all usages of the role set to the default of 10 years, except for the nginx tests where two certs are generated)
- update tomcat version

**Review by p-j-smith — DISMISSED** [review]

File: roles/ssl_certificates/README.md [hunk 1]
```diff
@@ -95,6 +95,7 @@ ssl_certificate:
   csr_common_name: "db"
```
Comment (drmatthews) [inline roles/ssl_certificates/README.md:95]:
```suggestion
  csr_common_name: db
```

**Review by drmatthews — COMMENTED** [review]
Looks great!

Very minor (and ignorable) but I removed some unnecessary double quotes from the example in the `README.md` file of the `nginx`.

**Review by drmatthews — APPROVED** [review]
🚀

**Comment by ruaridhg** [comment]
> @ruaridhg I've pushed some changes to:
> 
> * check certificate expiry date in the nginx role before copying to/from cache
> * added variable to `ssl_certificates` role to set the expiry date (all usages of the role set to the default of 10 years, except for the nginx tests where two certs are generated)
> * update tomcat version

Thanks, this looks like it covers everything we discussed!

**Pull request merged**